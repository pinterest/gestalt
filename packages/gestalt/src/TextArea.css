.textArea {
  composes: accessibilityOutline from "./Focus.css";
  composes: borderBox from "./Layout.css";
  composes: Text fontSize3 from "./Text.css";
  composes: leadingTall from "./Typography.css";
  composes: xsCol12 from "./Column.css";
  appearance: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  padding: 10px 14px;
  resize: none;
}

.textArea:focus {
  background-color: #fff;
}

.textArea::placeholder {
  color: #8e8e8e;
}

.normal {
  border-color: #8e8e8e;
}

textarea.errored {
  border-color: #e3780c;
}

textarea.errored:focus {
  box-shadow: 0 0 0 4px rgba(227, 120, 12, 0.75);
  outline: none;
}

.enabled {
  composes: darkGray from "./Colors.css";
  composes: whiteBg from "./Colors.css";
  cursor: text;
}

.disabled {
  composes: gray from "./Colors.css";
  composes: lightGrayBg from "./Colors.css";
}
