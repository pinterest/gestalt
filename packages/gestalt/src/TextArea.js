// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Box from './Box.js';
import FormErrorMessage from './FormErrorMessage.js';
import styles from './TextArea.css';

type Props = {|
  errorMessage?: string,
  disabled?: boolean,
  hasError?: boolean,
  id: string,
  name?: string,
  onBlur?: ({
    event: SyntheticFocusEvent<HTMLTextAreaElement>,
    value: string,
  }) => void,
  onChange: ({
    event: SyntheticInputEvent<HTMLTextAreaElement>,
    value: string,
  }) => void,
  onFocus?: ({
    event: SyntheticFocusEvent<HTMLTextAreaElement>,
    value: string,
  }) => void,
  onKeyDown?: ({
    event: SyntheticKeyboardEvent<HTMLTextAreaElement>,
    value: string,
  }) => void,
  placeholder?: string,
  rows?: number /* default: 3 */,
  value?: string,
|};

type State = {|
  focused: boolean,
|};

export default class TextArea extends React.Component<Props, State> {
  static propTypes = {
    disabled: PropTypes.bool,
    errorMessage: PropTypes.string,
    hasError: PropTypes.bool,
    id: PropTypes.string.isRequired,
    name: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onKeyDown: PropTypes.func,
    placeholder: PropTypes.string,
    rows: PropTypes.number,
    value: PropTypes.string,
  };

  static defaultProps = {
    disabled: false,
    hasError: false,
    rows: 3,
  };

  state = {
    focused: false,
  };

  setTextAreaRef = (ref: ?HTMLTextAreaElement) => {
    this.textarea = ref;
  };

  handleChange = (event: SyntheticInputEvent<HTMLTextAreaElement>) => {
    const { onChange } = this.props;
    onChange({ event, value: event.currentTarget.value });
  };

  handleBlur = (event: SyntheticFocusEvent<HTMLTextAreaElement>) => {
    const { onBlur } = this.props;
    if (onBlur) {
      onBlur({ event, value: event.currentTarget.value });
    }
  };

  handleFocus = (event: SyntheticFocusEvent<HTMLTextAreaElement>) => {
    const { onFocus } = this.props;
    if (onFocus) {
      onFocus({ event, value: event.currentTarget.value });
    }
  };

  handleKeyDown = (event: SyntheticKeyboardEvent<HTMLTextAreaElement>) => {
    const { onKeyDown } = this.props;
    if (onKeyDown) {
      onKeyDown({ event, value: event.currentTarget.value });
    }
  };

  // NOTE: we cannot move to React createRef until we audit uses of callsites
  // that reach into this component and use this instance variable
  textarea: ?HTMLElement;

  render() {
    const {
      disabled,
      errorMessage,
      hasError,
      id,
      name,
      placeholder,
      rows,
      value,
    } = this.props;

    const { focused } = this.state;

    const classes = classnames(
      styles.textArea,
      disabled ? styles.disabled : styles.enabled,
      hasError || errorMessage ? styles.errored : styles.normal
    );

    return (
      <span>
        <textarea
          aria-describedby={errorMessage && focused ? `${id}-error` : null}
          aria-invalid={errorMessage || hasError ? 'true' : 'false'}
          className={classes}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          onKeyDown={this.handleKeyDown}
          placeholder={placeholder}
          ref={this.setTextAreaRef}
          rows={rows}
          value={value}
        />
        {errorMessage && (
          <Box marginTop={1}>
            <FormErrorMessage id={id} text={errorMessage} />
          </Box>
        )}
      </span>
    );
  }
}
