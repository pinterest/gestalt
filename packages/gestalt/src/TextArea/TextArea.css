.textArea {
  composes: accessibilityOutline from "../Focus.css";
  composes: borderBox from "../Layout.css";
  composes: Text fontSize3 from "../Text/Text.css";
  composes: leadingTall from "../Typography.css";
  composes: xsCol12 from "../Column.css";
  appearance: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  padding: 10px 14px;
  resize: none;
}

.textArea:focus {
  background-color: #fff;
}

.textArea::placeholder {
  color: #b5b5b5;
}

.normal {
  border-color: #b5b5b5;
}

.errored {
  border-color: #e3780c;
}

.enabled {
  composes: darkGray from "../Colors.css";
  composes: whiteBg from "../Colors.css";
  cursor: text;
}

.disabled {
  composes: gray from "../Colors.css";
  composes: lightGrayBg from "../Colors.css";
}
