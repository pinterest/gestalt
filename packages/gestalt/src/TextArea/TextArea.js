// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import ErrorFlyout from '../ErrorFlyout/ErrorFlyout';
import styles from './TextArea.css';

type State = {
  focused: boolean,
  errorIsOpen: boolean,
};

type Props = {|
  errorMessage?: string,
  disabled?: boolean,
  id: string,
  idealErrorDirection?: 'up' | 'right' | 'down' | 'left' /* default: right */,
  name?: string,
  onBlur?: ({ event: SyntheticFocusEvent<>, value: string }) => void,
  onChange: ({ event: SyntheticInputEvent<>, value: string }) => void,
  onFocus?: ({ event: SyntheticFocusEvent<>, value: string }) => void,
  placeholder?: string,
  rows?: number /* default: 3 */,
  value?: string,
|};

export default class TextArea extends React.Component<Props, State> {
  static propTypes = {
    disabled: PropTypes.bool,
    errorMessage: PropTypes.string,
    id: PropTypes.string.isRequired,
    idealErrorDirection: PropTypes.string,
    name: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    placeholder: PropTypes.string,
    rows: PropTypes.number,
    value: PropTypes.string,
  };

  static defaultProps = {
    disabled: false,
    idealErrorDirection: 'right',
    rows: 3,
  };

  state: State = {
    focused: false,
    errorIsOpen: false,
  };

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.errorMessage !== this.props.errorMessage) {
      this.setState({ errorIsOpen: !!nextProps.errorMessage });
    }
  }

  textarea: ?HTMLElement;

  handleChange = (event: SyntheticInputEvent<>) => {
    if (event.target instanceof HTMLTextAreaElement) {
      this.props.onChange({
        event,
        value: event.target.value,
      });

      if (this.props.errorMessage) {
        this.setState({ errorIsOpen: true });
      }
    }
  };

  handleBlur = (event: SyntheticFocusEvent<>) => {
    if (this.props.errorMessage) {
      this.setState({ errorIsOpen: false });
    }
    if (event.target instanceof HTMLTextAreaElement && this.props.onBlur) {
      this.props.onBlur({
        event,
        value: event.target.value,
      });
    }
  };

  handleFocus = (event: SyntheticFocusEvent<>) => {
    if (this.props.errorMessage) {
      this.setState({ errorIsOpen: true });
    }
    if (event.target instanceof HTMLTextAreaElement && this.props.onFocus) {
      this.props.onFocus({
        event,
        value: event.target.value,
      });
    }
  };

  render() {
    const {
      disabled,
      errorMessage,
      id,
      idealErrorDirection,
      name,
      placeholder,
      rows,
      value,
    } = this.props;

    const classes = classnames(
      styles.textArea,
      disabled ? styles.disabled : styles.enabled,
      errorMessage ? styles.errored : styles.normal
    );

    return (
      <span>
        <textarea
          aria-describedby={
            errorMessage && this.state.focused ? `${id}-gestalt-error` : null
          }
          aria-invalid={errorMessage ? 'true' : 'false'}
          className={classes}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          placeholder={placeholder}
          ref={c => {
            this.textarea = c;
          }}
          rows={rows}
          value={value}
        />
        {errorMessage && this.state.errorIsOpen ? (
          <ErrorFlyout
            anchor={this.textarea}
            id={`${id}-gestalt-error`}
            idealDirection={idealErrorDirection}
            message={errorMessage}
            onDismiss={() => this.setState({ errorIsOpen: false })}
            size="sm"
          />
        ) : null}
      </span>
    );
  }
}
