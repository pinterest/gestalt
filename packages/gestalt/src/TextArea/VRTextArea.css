.input {
  appearance: none;
  background: none;
  border: none;
  color: inherit;
  display: block;
  font-size: var(--sema-font-size-ui-md);
  font-weight: var(--base-font-weight-400);
  line-height: 1.5;
  outline: 0;
  padding: 0;
  resize: none;
  width: 100%;
}

.inputParent {
  border-radius: var(--sema-rounding-300);
  box-sizing: border-box;
  display: block;
  isolation: isolate;
  position: relative;
  width: 100%;
}

.enabledText {
  color: var(--sema-color-text-default);
}

.disabledText {
  color: var(--sema-color-text-disabled);
}

.borderFocus {
  border: 1px solid var(--sema-color-border-focus-inner-default);
  outline: 2px solid var(--sema-color-border-focus-outer-default);
}

.enabled {
  background: var(--sema-color-background-default);
}

.enabledBorder {
  border: 1px solid var(--sema-color-border-interactive);
}

.enabledBorderHover {
  border: 1px solid var(--sema-color-hover-border-interactive);
}

.errorBorder {
  border: 1.5px solid var(--sema-color-border-error);
}

.errorBorderHover {
  border: 1.5px solid var(--sema-color-hover-border-error);
}

.errorBorderFocus {
  border: 1.5px solid var(--sema-color-border-focus-inner-default);
  outline: 2px solid var(--sema-color-border-focus-outer-default);
}

.disabled {
  background-color: var(--sema-color-background-disabled);
  border: 1px solid var(--sema-color-border-disabled);
}

.label {
  color: inherit;
  font-size: var(--sema-font-size-ui-xs);
  font-weight: var(--base-font-weight-500);
  line-height: var(sema-font- lineheight-UI-XS);
  padding: 0;
  position: absolute;
  z-index: 1;
}

.md_input {
  font-size: var(--sema-font-size-ui-md);
  min-height: 36px;
  padding-bottom: var(--sema-space-200);
}

.md_visibleLabel {
  padding-top: calc(
    var(--sema-space-300) + (var(--sema-font-size-ui-md) * 1.1)
  );
}

.md_noLabel {
  padding-top: var(--sema-space-200);
}

html[dir="rtl"] .md_inputHorizontalPadding {
  padding-left: var(--sema-space-200);
  padding-right: var(--sema-space-300);
}

html:not([dir="rtl"]) .md_inputHorizontalPadding {
  padding-left: var(--sema-space-300);
  padding-right: var(--sema-space-200);
}

html[dir="rtl"] .md_label {
  left: calc(var(--sema-space-300));
  right: calc(var(--sema-space-300) + 1px);
}

html:not([dir="rtl"]) .md_label {
  left: calc(var(--sema-space-300) + 1px);
  right: calc(var(--sema-space-300));
}

.md_labelPos {
  top: var(--sema-space-200);
}
