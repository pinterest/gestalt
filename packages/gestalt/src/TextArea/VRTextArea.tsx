import { forwardRef, Fragment, ReactNode, useImperativeHandle, useRef, useState } from 'react';
import classnames from 'classnames';
import styles from './VRTextArea.css';
import boxStyles from '../Box.css';
import FormErrorMessage from '../sharedSubcomponents/FormErrorMessage';
import FormHelperText from '../sharedSubcomponents/FormHelperText';
import { MaxLength } from '../TextField';
import typographyStyle from '../Typography.css';
import useInteractiveStates from '../utils/useInteractiveStates';

type Props = {
  // REQUIRED
  id: string;
  onChange: (arg1: { event: React.ChangeEvent<HTMLTextAreaElement>; value: string }) => void;
  // OPTIONAL
  dataTestId?: string;
  disabled?: boolean;
  errorMessage?: ReactNode;
  hasError?: boolean;
  helperText?: string;
  label?: string;
  labelDisplay?: 'visible' | 'hidden';
  maxLength?: MaxLength | null | undefined;
  name?: string;
  onBlur?: (arg1: { event: React.FocusEvent<HTMLTextAreaElement>; value: string }) => void;
  onClick?: (arg1: { event: React.MouseEvent<HTMLTextAreaElement>; value: string }) => void;
  onFocus?: (arg1: { event: React.FocusEvent<HTMLTextAreaElement>; value: string }) => void;
  onKeyDown?: (arg1: { event: React.KeyboardEvent<HTMLTextAreaElement>; value: string }) => void;
  placeholder?: string;
  readOnly?: boolean;
  rows?: number;
  value?: string;
};

const TextAreaWithForwardRef = forwardRef<HTMLTextAreaElement, Props>(function TextArea(
  {
    dataTestId,
    disabled = false,
    errorMessage,
    hasError = false,
    helperText,
    id,
    label,
    labelDisplay,
    maxLength,
    name,
    onBlur,
    onChange,
    onClick,
    onFocus,
    onKeyDown,
    placeholder,
    readOnly,
    rows = 3,
    value,
  }: Props,
  ref,
) {
  const innerRef = useRef<null | HTMLTextAreaElement>(null);
  const labelRef = useRef<null | HTMLLabelElement>(null);

  const {
    handleOnBlur,
    handleOnFocus,
    handleOnMouseEnter,
    handleOnMouseLeave,
    isHovered,
    isFocused,
  } = useInteractiveStates();

  // @ts-expect-error - TS2322 - Type 'HTMLDivElement | HTMLTextAreaElement | null' is not assignable to type 'HTMLTextAreaElement'.
  useImperativeHandle(ref, () => innerRef.current);

  const hasErrorMessage = Boolean(errorMessage);

  const isLabelVisible = labelDisplay === 'visible';

  // ==== STATE ====
  const [currentLength, setCurrentLength] = useState(value?.length ?? 0);

  // ==== A11Y ====

  let ariaDescribedby;

  if (hasErrorMessage) {
    ariaDescribedby = `${id}-error`;
  }

  if (helperText || maxLength) {
    ariaDescribedby = `${id}-helperText`;
  }

  return (
    <Fragment>
      <div onMouseEnter={handleOnMouseEnter} onMouseLeave={handleOnMouseLeave}>
        <div
          className={classnames(
            styles.inputParent,
            styles.md_input,
            styles.md_inputHorizontalPadding,
            {
              [styles.disabled]: disabled,
              [styles.disabledText]: disabled,
              [styles.enabled]: !disabled,
              [styles.enabledText]: !disabled,
              [styles.borderFocus]: !disabled && isFocused && !hasErrorMessage,
              [styles.errorBorderFocus]: !disabled && isFocused && hasErrorMessage,
              [styles.enabledBorder]: !disabled && !isFocused && !hasErrorMessage && !isHovered,
              [styles.enabledBorderHover]: !disabled && !isFocused && !hasErrorMessage && isHovered,
              [styles.errorBorder]: !disabled && !isFocused && hasErrorMessage,
              [styles.errorBorderHover]: !disabled && !isFocused && hasErrorMessage && isHovered,
              // md
              [styles.md_visibleLabel]: label && isLabelVisible,
              [styles.md_noLabel]: !label || (label && !isLabelVisible),
            },
          )}
        >
          {label && (
            <label
              ref={labelRef}
              className={classnames(
                styles.label,
                styles.md_label,
                styles.md_labelPos,
                typographyStyle.truncate,
                {
                  [boxStyles.visuallyHidden]: !isLabelVisible,
                },
              )}
              htmlFor={id}
            >
              {label}
            </label>
          )}
          <textarea
            ref={innerRef}
            aria-describedby={isFocused ? ariaDescribedby : undefined}
            aria-invalid={hasErrorMessage || hasError ? 'true' : 'false'}
            className={classnames(styles.input)}
            data-test-id={dataTestId}
            disabled={disabled}
            id={id}
            maxLength={maxLength?.characterCount}
            name={name}
            onBlur={(event) => {
              handleOnBlur();
              onBlur?.({ event, value: event.currentTarget.value });
            }}
            onChange={(event) => {
              setCurrentLength(event.currentTarget.value?.length ?? 0);
              onChange({ event, value: event.currentTarget.value });
            }}
            onClick={(event) => onClick?.({ event, value: event.currentTarget.value })}
            onFocus={(event) => {
              handleOnFocus();
              onFocus?.({ event, value: event.currentTarget.value });
            }}
            onKeyDown={(event) => onKeyDown?.({ event, value: event.currentTarget.value })}
            placeholder={placeholder}
            readOnly={readOnly}
            rows={rows}
            value={value}
          />
        </div>
      </div>
      {(helperText || maxLength) && !hasErrorMessage ? (
        <FormHelperText
          currentLength={currentLength}
          disabled={disabled}
          id={`${id}-helperText`}
          maxLength={maxLength}
          size="md"
          text={helperText}
        />
      ) : null}

      {!disabled && hasErrorMessage ? (
        <FormErrorMessage id={`${id}-error`} size="md" text={errorMessage} />
      ) : null}
    </Fragment>
  );
});

TextAreaWithForwardRef.displayName = 'TextArea';

export default TextAreaWithForwardRef;
