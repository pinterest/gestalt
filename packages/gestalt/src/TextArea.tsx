import { forwardRef, ReactElement, ReactNode, useState } from 'react';
import classnames from 'classnames';
import Box from './Box';
import focusStyles from './Focus.css';
import formElement from './sharedSubcomponents/FormElement.css';
import FormErrorMessage from './sharedSubcomponents/FormErrorMessage';
import FormHelperText from './sharedSubcomponents/FormHelperText';
import FormLabel from './sharedSubcomponents/FormLabel';
import TagArea from './TagArea/TagArea';
import styles from './TextArea.css';
import VRTextArea from './TextArea/VRTextArea';
import useInExperiment from './useInExperiment';

const ROW_HEIGHT = 24;
const INPUT_PADDING_WITH_TAGS = 20;

type Props = {
  /**
   * Indicate if the input is currently disabled. See the [disabled example](https://gestalt.pinterest.systems/web/textarea#Disabled) for more details.
   */
  disabled?: boolean;
  /**
   * Available for testing purposes, if needed. Consider [better queries](https://testing-library.com/docs/queries/about/#priority) before using this prop.
   */
  dataTestId?: string;
  /**
   * For most use cases, pass a string with a helpful error message (be sure to localize!). In certain instances it can be useful to make some text clickable; to support this, you may instead pass a React.Node to wrap text in Link or TapArea. See the [error message example](https://gestalt.pinterest.systems/web/textarea#Error-message) for more details.
   */
  errorMessage?: ReactNode;
  /**
   * This field is deprecated and will be removed soon. Please do not use.
   */
  hasError?: boolean;
  /**
   * More information about how to complete the form field. See the [helper text example](https://gestalt.pinterest.systems/web/textarea#Helper-text) for more details.
   */
  helperText?: string;
  /**
   * A unique identifier for the input.
   */
  id: string;
  /**
   * The label for the input. Be sure to localize the text.
   */
  label?: string;
  /**
   * Whether the label should be visible or not. If `hidden`, the label is still available for screen reader users, but does not appear visually. See the [label visibility variant](https://gestalt.pinterest.systems/web/textarea#Label-visibility) for more info.
   */
  labelDisplay?: 'visible' | 'hidden';
  /**
   * The maximum number of characters allowed in TextArea. `maxLength` must be an integer value 0 or higher. See the [maximum length variant](https://gestalt.pinterest.systems/web/textarea#Maximum-length) for more details.
   */
  maxLength?: {
    characterCount: number;
    errorAccessibilityLabel: string;
  };
  /**
   * A unique name for the input.
   */
  name?: string;
  /**
   * Callback triggered when the user blurs the input.!
   */
  onBlur?: (arg1: { event: React.FocusEvent<HTMLTextAreaElement>; value: string }) => void;
  /**
   * Callback triggered when the value of the input changes.
   */
  onChange: (arg1: { event: React.ChangeEvent<HTMLTextAreaElement>; value: string }) => void;
  /**
   * Callback triggered when the user focuses the input.
   */
  onFocus?: (arg1: { event: React.FocusEvent<HTMLTextAreaElement>; value: string }) => void;
  /**
   * Callback triggered when the user presses any key while the input is focused.
   */
  onKeyDown?: (arg1: { event: React.KeyboardEvent<HTMLTextAreaElement>; value: string }) => void;
  /**
   * Placeholder text shown the the user has not yet input a value.
   */
  placeholder?: string;
  /**
   * Indicate if the input is currently readOnly. See the [readOnly example](https://gestalt.pinterest.systems/web/textarea#Read-only) for more details.
   */
  readOnly?: boolean;
  /**
   * Ref that is forwarded to the underlying input element. See the [ref example](https://gestalt.pinterest.systems/web/textarea#With-a-ref) for more details.
   */
  ref?: ReactElement; // eslint-disable-line react/no-unused-prop-types,
  /**
   * Number of text rows to display. Note that tags take up more space, and will show fewer rows than specified.
   */
  rows?: number;
  /**
   * List of tags to display in the component. See the [tags example](https://gestalt.pinterest.systems/web/textarea#With-tags) for more details.
   */
  tags?: ReadonlyArray<ReactElement>;
  /**
   * The current value of the input.
   */
  value?: string;
};

/**
 * [TextArea](https://gestalt.pinterest.systems/web/textarea) allows for multi-line input.
 *
 * ![TextArea light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/TextArea.spec.ts-snapshots/TextArea-chromium-darwin.png)
 * ![TextArea dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/TextArea-dark.spec.ts-snapshots/TextArea-dark-chromium-darwin.png)
 *
 */

const TextAreaWithForwardRef = forwardRef<HTMLTextAreaElement, Props>(function TextArea(
  {
    dataTestId,
    disabled = false,
    errorMessage,
    hasError = false,
    helperText,
    id,
    label,
    labelDisplay = 'visible',
    maxLength,
    name,
    onBlur,
    onChange,
    onFocus,
    onKeyDown,
    placeholder,
    readOnly = false,
    rows = 3,
    tags,
    value,
  }: Props,
  ref,
) {
  const [focused, setFocused] = useState(false);
  const [currentLength, setCurrentLength] = useState(value?.length ?? 0);

  const isInVRExperiment = useInExperiment({
    webExperimentName: 'web_gestalt_visualRefresh',
    mwebExperimentName: 'web_gestalt_visualRefresh',
  });

  const handleChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setCurrentLength(event.currentTarget.value?.length ?? 0);
    onChange({ event, value: event.currentTarget.value });
  };

  const handleBlur = (event: React.FocusEvent<HTMLTextAreaElement>) => {
    setFocused(false);
    if (onBlur) {
      onBlur({ event, value: event.currentTarget.value });
    }
  };

  const handleFocus = (event: React.FocusEvent<HTMLTextAreaElement>) => {
    setFocused(true);
    if (onFocus) {
      onFocus({ event, value: event.currentTarget.value });
    }
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLTextAreaElement>) => {
    if (onKeyDown) {
      onKeyDown({ event, value: event.currentTarget.value });
    }
  };

  const hasErrorMessage = Boolean(errorMessage);

  const classes = classnames(
    styles.textArea,
    formElement.base,
    formElement.lg,
    disabled ? formElement.disabled : formElement.enabled,
    (hasError || hasErrorMessage) && !focused ? formElement.errored : formElement.normal,
    tags
      ? {
          [focusStyles.accessibilityOutlineFocus]: focused,
          [styles.textAreaWrapper]: true,
        }
      : {},
  );

  if (maxLength && maxLength.characterCount < 0) {
    throw new Error('`maxLength` must be an integer value 0 or higher.');
  }

  let ariaDescribedby;

  if (hasErrorMessage) {
    ariaDescribedby = `${id}-error`;
  }

  if (helperText || maxLength) {
    ariaDescribedby = `${id}-helperText`;
  }

  const inputElement = (
    <textarea
      // checking for "focused" is not required by screenreaders but it prevents a11y integration tests to complain about missing label, as aria-describedby seems to shadow label in tests though it's a W3 accepeted pattern https://www.w3.org/TR/WCAG20-TECHS/ARIA1.html
      ref={ref}
      aria-describedby={focused ? ariaDescribedby : undefined}
      aria-invalid={hasErrorMessage || hasError ? 'true' : 'false'}
      className={tags ? styles.unstyledTextArea : classes}
      data-test-id={dataTestId}
      disabled={disabled}
      id={id}
      maxLength={maxLength?.characterCount}
      name={name}
      onBlur={handleBlur}
      onChange={handleChange}
      onFocus={handleFocus}
      onKeyDown={handleKeyDown}
      placeholder={placeholder}
      readOnly={readOnly}
      rows={tags ? undefined : rows}
      value={value}
    />
  );

  const tagsWrapperStyle = {
    minHeight: rows * ROW_HEIGHT + INPUT_PADDING_WITH_TAGS,
    maxHeight: rows * ROW_HEIGHT + INPUT_PADDING_WITH_TAGS,
  } as const;

  if (isInVRExperiment && !tags)
    return (
      <VRTextArea
        ref={ref}
        dataTestId={dataTestId}
        disabled={disabled}
        errorMessage={errorMessage}
        hasError={hasError}
        helperText={helperText}
        id={id}
        label={label}
        labelDisplay={labelDisplay}
        maxLength={maxLength}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        placeholder={placeholder}
        readOnly={readOnly}
        rows={rows}
        value={value}
      />
    );

  if (isInVRExperiment && tags)
    return (
      <TagArea
        ref={ref}
        dataTestId={dataTestId}
        disabled={disabled}
        errorMessage={errorMessage}
        hasError={hasError}
        helperText={helperText}
        id={id}
        label={label}
        labelDisplay={labelDisplay}
        maxLength={maxLength}
        name={name}
        // @ts-expect-error - TS2322
        onBlur={onBlur}
        // @ts-expect-error - TS2322
        onChange={onChange}
        // @ts-expect-error - TS2322
        onFocus={onFocus}
        // @ts-expect-error - TS2322
        onKeyDown={onKeyDown}
        placeholder={placeholder}
        readOnly={readOnly}
        rows={rows}
        size="md"
        tags={tags}
        value={value}
      />
    );

  return (
    <span>
      {label && <FormLabel id={id} label={label} labelDisplay={labelDisplay} size="lg" />}
      {tags ? (
        <div className={classes} style={tagsWrapperStyle}>
          {tags.map((tag, tagIndex) => (
            // eslint-disable-next-line react/no-array-index-key
            <Box key={tagIndex} marginBottom={1} marginEnd={1}>
              {tag}
            </Box>
          ))}
          <Box flex="grow" maxWidth="100%" overflow="hidden" position="relative">
            {/* This is an invisible spacer div which mirrors the input's
             * content. We use it to implement the flex wrapping behavior
             * which is not supported by inputs, by having the actual input
             * track it with absolute positioning. */}
            <div aria-hidden className={styles.textAreaSpacer}>
              {/* Keep a non-empty value so that the div doesn't collapse */}
              {value || '-'}
            </div>
            {inputElement}
          </Box>
        </div>
      ) : (
        inputElement
      )}
      {(helperText || maxLength) && !errorMessage ? (
        <FormHelperText
          currentLength={currentLength}
          id={`${id}-helperText`}
          maxLength={maxLength}
          text={helperText}
        />
      ) : null}
      {hasErrorMessage && <FormErrorMessage id={`${id}-error`} text={errorMessage} />}
    </span>
  );
});

TextAreaWithForwardRef.displayName = 'TextArea';

export default TextAreaWithForwardRef;
