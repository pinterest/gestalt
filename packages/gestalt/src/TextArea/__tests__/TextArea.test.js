/* eslint-env jest */
import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import TextArea from '../TextArea';

describe('TextArea', () => {
  it('Renders an ErrorFlyout if an error message is passed in', () => {
    const wrapper = shallow(
      <TextArea errorMessage="test" id="test" onChange={jest.fn()} />
    );
    wrapper.instance().setState({ errorIsOpen: true });
    wrapper.simulate('focus');
    expect(wrapper.find('ErrorFlyout').length).toEqual(1);
  });

  it('Does not render an ErrorFlyout when errorMessage is null', () => {
    const wrapper = shallow(<TextArea id="test" onChange={jest.fn()} />);
    expect(wrapper.find('ErrorFlyout').length).toEqual(0);
  });

  it('TextArea normal', () => {
    const tree = create(
      <TextArea
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('TextArea with error', () => {
    const tree = shallow(
      <TextArea
        errorMessage="error message"
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });

  it('TextArea with rows', () => {
    const tree = create(
      <TextArea id="test" onChange={jest.fn()} rows={5} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('TextArea with errorMessage prop change', () => {
    const tree = shallow(
      <TextArea
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    );
    expect(tree.find('ErrorFlyout').length).toEqual(0);
    tree.setProps({
      errorMessage: 'error message',
    });
    expect(tree.find('ErrorFlyout').length).toEqual(1);
  });
});
