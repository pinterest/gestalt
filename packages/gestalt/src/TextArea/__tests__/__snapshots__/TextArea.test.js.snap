// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea TextArea normal 1`] = `
<span>
  <textarea
    aria-describedby={null}
    aria-invalid="false"
    className="textArea enabled normal"
    disabled={false}
    id="test"
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    rows={3}
    value={undefined}
  />
</span>
`;

exports[`TextArea TextArea with error 1`] = `"<span><textarea aria-invalid=\\"true\\" class=\\"textArea enabled errored\\" id=\\"test\\" rows=\\"3\\"></textarea></span>"`;

exports[`TextArea TextArea with rows 1`] = `
<span>
  <textarea
    aria-describedby={null}
    aria-invalid="false"
    className="textArea enabled normal"
    disabled={false}
    id="test"
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    rows={5}
    value={undefined}
  />
</span>
`;
